### PR TITLE
Introduce CLI for ApplySet-based pruning

### PR DESCRIPTION
**Pull request by KnVerey**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is the first of a series of PRs to implement the ApplySet KEP in kubectl: git.k8s.io/enhancements/keps/sig-cli/3659-kubectl-apply-prune. Justin and I would like to merge our work in smaller chunks like this for ease of iteration and review. We expect to have the full alpha feature complete in time for code freeze.

Specifically, this PR wires up the `--applyset` flag on the `apply` command, subject to an alpha environment variable gate `KUBECTL_APPLYSET=true`. The code currently has two extra places that return "not implemented" errors, which will be removed as the feature progresses.

Here is the relevant section of the `kubectly apply -h` help text with the env variable enabled:

```sh
    --applyset='':
	[alpha] The name of the ApplySet that tracks which resources are being managed, for the purposes of
	determining what to prune. Live resources that are part of the ApplySet but have been removed from the
	provided configs will be deleted. Format: [RESOURCE][.GROUP]/NAME. A Secret will be used if no resource or
	group is specified.

    --prune=false:
	Automatically delete previously applied resource objects that do not appear in the provided configs. For
	alpha1, use with either -l or --all. For alpha2, use with --applyset.
```

#### Special notes for your reviewer:

Are there any other flag combinations we should exclude?

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

(This will certainly be user-facing, but we will include the release note on the future PR that actually enables the feature)

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/42a4b1c8a18780cf1d5c2460113d59b246002548/keps/sig-cli/3659-kubectl-apply-prune
```

### CODE REVIEW

**Comment by KnVerey** [comment]
/hold
for feedback from an approver other than myself

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:91]: The change above is not at all related to this PR. I was searching the file for examples of how to test flag exclusion, and this one caught my eye. In reading it, I realized it doesn't prove anything! It was just testing that the helper (which I deleted below) returns an error, which it always does. Since `ToOptions` seems responsible for extra args errors in the real command, I wired that up in the test.

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:84]: This is also drive-by clean up. After digging around trying to find a pattern for enabling and testing the alpha enablement env vars that we always require in kubectl and coming up empty, I decided to implement what I was looking for, and put it to use where currently relevant. Same for the test helper I added. LMK if I missed tools I should use instead (or if there are other alphas I should modify as well).

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:984]: Nit: if we don't already have a feature flag type and we wanted to be a bit more type-safe, we could do something like:

```
type FeatureFlag string
var FeatureFlag ApplySetEnv = "FOO"

func (f *FeatureFlag) IsEnabled() {
...
```

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:122]: Maybe should be a `[]string`, but I also think it's reasonable to wait until we need to set more than one

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:49]: My 2c is that extracting this to a constant obscures more than it helps, but I know others disagree.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:27]: Super-nit: personally I would make this a GVR, even though it would mean not being able to keep it as `const`.

File: staging/src/k8s.io/kubectl/pkg/cmd/testing/util.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/testing/util.go:184]: Nit: I suggest either adding a comment here, or using a strongly-typed value for key, or renaming to RunWithEnvVar.  I know it's test code, but it's otherwise what this function does is a little obscure (to me, anyway).

**Review by justinsb — COMMENTED** [review]
A few nits, but overall it's behind a feature flag and improves testing, so lgtm

**Comment by justinsb** [comment]
A few nits / ideas for the future, but as this is behind a well-tested feature flag:

/lgtm

(Has hold anyway, so this won't accidentally cause a merge)

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:984]: Love this suggestion 👌

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:49]: I don't feel strongly either way tbh

**Comment by KnVerey** [comment]
/label tide/merge-method-squash

I like your suggestions--updated!

**Comment by KnVerey** [comment]
> This PR does enable the feature, then (albeit gated)?

Yes, sort of... but it does nothing but throw an error. This is what someone would see if they try it:

```
KUBECTL_APPLYSET=true kubectl apply --prune --applyset=foo -f ~/scratch/resources.yaml
error: --applyset is not yet supported
```

**Comment by justinsb** [comment]
Thanks for updates, updates lgtm so reapplying (but we still have hold)

/lgtm

And re @sftim I am a proponent of doing things this way - I think breaking things into smaller PRs allows for better PR review and more effective course correction if people disagree with the direction.  In the KEP we did describe how important user feedback is prior to GA; my 2c is that we should be OK with users that run prerelease code/binaries _and_ enable a feature flag getting a half-baked feature.  Right now the behaviour users can unlock is an error message, which seems very safe.  We're planning on having this in a user-acceptable state by feature-freeze, but if that doesn't go to plan we'll likely just keep/reintroduce the error message.  We could alternatively just send an all-in-one PR chain, but I don't (personally) see an advantage.

That said, if there's a flag or significant change that is visible _without_ the feature-flag, that would be something we would definitely want to fix.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (justinsb) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:55]: At some stage we might want two separate methods - one which deals with the flags, and one is explicit.  The more we can avoid pushing knowledge of flags down deep into the logic, the simpler it will be IMO.  That said, no need for the second method yet!

**Comment by justinsb** [comment]
Thanks for adding the logic around namespace!

/lgtm

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:304]: What is the reason of this?

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:28]: If we change this to something different(e.g. configmaps), what will happen when user upgrades their cluster?, will this be a breaking change?

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:39]: We don't support pruning resources in default namespace?

File: staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:432]: Awesome to manage all environment variables in one place.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:229]: Nit: I wouldn't consider `--all` a prune-only flag, so I'd suggest to leave it here.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:94]: Nit: I prefer not to test exact error messages, this saves time when they change during other updates. I'd focus only on testing substrings which confirm what you're testing.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:39]: default or any other is just a name like any other :wink:

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:28]: Yes, that will be a breaking change, but this is alpha for now, so I'm not worried about such aspect.

**Review by soltysh — APPROVED** [review]
Minor nits that can be addressed in planned followups, but this is good to go as is.
/lgtm
/approve

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:229]: I know that `--selector` affects the main `apply` operation, but I thought that `--all` only affects pruning. I found that it was definitely added for pruning: https://github.com/kubernetes/kubernetes/commit/56a22f925f6f1fd774ad1ae9e04bcf8d75bbde63#diff-e255643fe2dc451729fc438dad24902dff489a8801a5c976c2d783649ceb870bR103. What else does it do now? Based on the code, it is only used during flag validation, to ensure the user intended to target everything if they omitted `--selector`.

That PR also shows that `--grace-period` and `--cascade` were also added for pruning originally, but based on the coded I believe that those also affect the `--force` flag these days.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:304]: It is to be very explicit for everything related to the pruning operation, and to err on the side of being more restrictive to start and relaxing later if needed. Note that 'enforcing' the namespace on an object works the same way: if your object has no namespace in the config, one will be filled in based on the value of the namespace flag, but not the kubeconfig's current or default values.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:94]: I like that testing the whole message validates the real user experience we're providing on errors, and catches things like redundancy in the wrapping. But it's a tradeoff for sure, and in particular I buy the argument that the 'see -h' part isn't useful here. 😄

**Pull request merged**
